### PR TITLE
fix(test-runner): WebServer > Use socket for reuseExistingServer to detect servers that aren't using SO_REUSEADDR

### LINKED ISSUE #8513: [BUG] `webServer.reuseExistingServer` doesn't work with servers bound to a specific address
**Context:**
- Playwright Version: 1.14
- Operating System: Mac
- Node.js version: 16.8.0
- Browser: N/A
- Extra: N/A

**Code Snippet**
```javascript
const net = require('net');

const existingServer = net.createServer();
existingServer.listen(3000, '127.0.0.1', () => {
  canBindPort(3000).then((result) => {
    console.log(`Can bind port: ${result}`);
    existingServer.close();
  });
});

async function canBindPort(port) {
  return new Promise((resolve) => {
    const server = net.createServer();

    server.on('error', () => resolve(false));
    server.listen(port, () => {
      server.close(() => {
        resolve(true);
      });
    });
  });
}
```

**Describe the bug**

Above is a copy of the `canBindPort` function from `node_modules/@playwright/test/lib/test/webServer.js`.
The script starts a server that listens on `127.0.0.1:3000`, then calls `canBindPort(3000)`, which returns `true` when it shouldn't. The function only returns `false` if the `existingServer` is started without passing a specific address.

Found this while using the [`Vite`](https://vitejs.dev/) dev server.

### PR DESCRIPTION
**Pull request by FokkeZB**
Fixes #8513

There are two potential cause that I'm aware for the above linked bug:

1. A server is running on a specific local host (e.g. 127.0.0.1) while the test assumes none (defaulting to `locahost`).
2. A server is running that doesn't use `SO_REUSEADDR`.

This PR fixes both causes by using a socket connection to test if the port is in use, instead of trying to start a server on the port. Funny enough this was exactly what the code was already doing _after_ starting the server, to check if it is up already.

I still like the suggestion made to make the host configurable, but for other reasons, which is that it would indirectly set `baseURL` to meet the user's expectation better.

### CODE REVIEW

File: src/test/webServer.ts [hunk 1]
```diff
@@ -103,31 +103,23 @@ export class WebServer {
   }
 }
 
-async function canBindPort(port: number): Promise<boolean> {
+async function isPortUsed(port: number): Promise<boolean> {
```
Comment (FokkeZB) [inline src/test/webServer.ts:106]: Replaced `canBindPort` by `isPortUsed` using the code in `waitForSocket`, which now uses this function along with the above private `WebServer._startProcess`.

**Comment by mxschmitt** [comment]
This looks good!

Regarding tests, see here: https://github.com/microsoft/playwright/blob/385d489b35ad5bc3b28c78c05b9c04e87400bd22/tests/playwright-test/web-server.spec.ts

As a command I'd pass a small Node.js script which does `process.exit(1)` since it never should get executed. And then create a test where you loop over `127.0.0.1` and `0.0.0.0` to ensure both is working. The actual servers you start inside the tests and stop it at the end of the test.

**Comment by mxschmitt** [comment]
@FokkeZB feel free to use the following test:

````ts
for (const host of ['localhost', '127.0.0.1', '0.0.0.0']) {
  test.only(`should detect the server if a web-server is already running on ${host}`, async ({ runInlineTest }, { workerIndex }) => {
    const port = workerIndex + 10500;
    const server = http.createServer((req: http.IncomingMessage, res: http.ServerResponse) => {
      res.end('<html><body>hello</body></html>');
    });
    await new Promise(resolve => server.listen(port, host, resolve));
    try {
      const result = await runInlineTest({
        'test.spec.ts': `
      const { test } = pwt;
      test('connect to the server via the baseURL', async ({baseURL, page}) => {
        await page.goto('/hello');
        expect(await page.textContent('body')).toBe('hello');
      });
    `,
        'playwright.config.ts': `
      module.exports = {
        webServer: {
          command: 'node -e "process.exit(1)"',
          port: ${port},
          reuseExistingServer: false,
        }
      };
    `,
      });
      expect(result.exitCode).toBe(1);
      expect(result.output).toContain(`Port ${port} is used, make sure that nothing is running on the port`);
    } finally {
      await new Promise(resolve => server.close(resolve));
    }
  });
}
```

**Comment by FokkeZB** [comment]
Thanks @mxschmitt, I've included the test.

**Review by mxschmitt — APPROVED** [review]

File: tests/playwright-test/web-server.spec.ts [hunk 2]
```diff
@@ -225,3 +225,37 @@ test('should throw when a server is already running on the given port and strict
   expect(result.output).toContain(`Port ${port} is used, make sure that nothing is running on the port`);
   await new Promise(resolve => server.close(resolve));
 });
+
+for (const host of ['localhost', '127.0.0.1', '0.0.0.0']) {
+  test.only(`should detect the server if a web-server is already running on ${host}`, async ({ runInlineTest }, { workerIndex }) => {
```
[the remark this answers is not in the document]
Comment (FokkeZB) [reply tests/playwright-test/web-server.spec.ts]: A shoot, missed that.

**Review by JoelEinbinder — APPROVED** [review]

**Pull request merged**